file/BinaryFileSource.cpp

```diff
@@ -38,102 +76,234 @@
  * |setter setFilePath(path)
  * |setter setAutoRewind(rewind)
  **********************************************************************/
-class BinaryFileSource : public Pothos::Block
+class BinaryFileSourceBase : public Pothos::Block
 {
 public:
-    static Block *make(const Pothos::DType &dtype)
-    {
-        return new BinaryFileSource(dtype);
-    }
-
-    BinaryFileSource(const Pothos::DType &dtype):
-        _fd(-1),
-        _rewind(false)
+    BinaryFileSourceBase(const Pothos::DType& dtype):
+        _path(),
+        _fileResourceMutex()
     {
         this->setupOutput(0, dtype);
-        this->registerCall(this, POTHOS_FCN_TUPLE(BinaryFileSource, setFilePath));
-        this->registerCall(this, POTHOS_FCN_TUPLE(BinaryFileSource, setAutoRewind));
+
+        this->registerCall(this, POTHOS_FCN_TUPLE(BinaryFileSourceBase, setFilePath));
+        this->registerCall(this, POTHOS_FCN_TUPLE(BinaryFileSourceBase, setAutoRewind));
     }
 
-    void setFilePath(const std::string &path)
+    virtual ~BinaryFileSourceBase() = default;
+
+    void setFilePath(const std::string& path)
     {
-        _path = path;
-        //file was open -> close old fd, and open this new path
-        if (_fd != -1)
+        if (!Poco::File(path).exists())
         {
-            this->deactivate();
-            this->activate();
+            throw Pothos::FileNotFoundException(path);
         }
+
+        _path = path;
+        this->deactivate();
+        this->activate();
     }
 
-    void setAutoRewind(const bool rewind)
+    virtual void setAutoRewind(const bool rewind)
     {
-        _rewind = rewind;
+        throw Pothos::AssertionViolationException();
+    }
+
+    // Make sure we only expose classes that implement these.
+    virtual void activate() = 0;
+    virtual void deactivate() = 0;
+    virtual void work() = 0;
+
+protected:
+    std::string _path;
+
+    Poco::FastMutex _fileResourceMutex;
+};
+
+class BinaryFileSource : public BinaryFileSourceBase
+{
+public:
+    static Block* make(const Pothos::DType& dtype)
+    {
+        return new BinaryFileSource(dtype);
+    }
+
+    BinaryFileSource(const Pothos::DType& dtype) :
+        BinaryFileSourceBase(dtype),
+        _fd(-1)
+    {
+    }
+
+    void setAutoRewind(const bool) override
+    {
+        throw Pothos::NotImplementedException("You must set optimizeForStandardFile to true to enable auto-rewind.");
     }
 
-    void activate(void)
+    void activate() override
     {
-        if (_path.empty()) throw Pothos::FileException("BinaryFileSource", "empty file path");
+        Poco::FastMutex::ScopedLock lock(_fileResourceMutex);
+
         _fd = open(_path.c_str(), O_RDONLY | O_BINARY);
-        if (_fd < 0)
-        {
-            poco_error_f4(Poco::Logger::get("BinaryFileSource"), "open(%s) returned %d -- %s(%d)", _path, _fd, std::string(strerror(errno)), errno);
-        }
+        if (_fd < 0) throw Pothos::Util::ErrnoException<Pothos::OpenFileException>();
```

Comment:
activate()/deactivate()/and work() should all be mutually exclusive. The "actor" should only allow one to be called at a time when he gets thread context. Did you see something weird here? Otherwise you shouldnt need the lock, the framework gives you thread safety

Reply:
I'll remove that then. I'm still figuring out this weird PothosFlow-specific crash, which I'm guessing has to do with call order, so I'll mark this as dontmerge for now.